src/main/scala/async/Listener.scala

```diff
@@ -1,73 +1,171 @@
     val nextNumber: Long
+    /** Attempt to lock the next lock. */
     def lockNext(): LockResult
 
-  type ReleaseBoundary = SemiLock | Locked.type
+  /** Points to a position on the lock chain, whose lock up until this point has been acquired,
+    * but no further.
+    */
+  type LockMarker = PartialLock | Locked.type
 
-  trait TopLock:
+  /** A lock required by a listener to be acquired before accepting values.
+    * Should there be multiple listeners that needs to be locked at the same time,
+    * they should be locked by larger-number-first.
+
+    * Some implementations are provided for ease of implementations:
+    * - For custom listener implementations involving locks: [[NumberedLock]] provides uniquely
+    * numbered locks.
+    * - For source transformation implementations: [[ListenerLockWrapper]] provides a ListenerLock instance
+    *   that only forwards the requests to the underlying lock. [[withLock]] is a convenient
+    *   `.map` for `[[ListenerLock]] | Null`.
+    */
+  trait ListenerLock:
+    /** The assigned number of the lock.
+      * If the listener holds inner listeners underneath that utilizes locks,
+      * it is **required** that [[selfNumber]] must be larger than any [[PartialLock.nextNumber]] of
+      * any returned [[PartialLock]]s.
+      */
     val selfNumber: Long
+    /** Attempt to lock the current [[ListenerLock]]. To try to lock all possible nesting locks,
+      * see [[Listener.lockCompletely]].
+      * Locks are guaranteed to be held as short as possible.
+      */
     def lockSelf(source: Async.Source[?]): LockResult
 
-  // TODO this should only be used for adapting the source
-  inline def mappingTopLock(inner: TopLock | Null, inline body: (Async.Source[?]) => LockResult) =
-    if inner != null then
-      new TopLock:
-        val selfNumber = inner.selfNumber
-        def lockSelf(source: Source[?]) = body(source)
+  /** A special wrapper for [[ListenerLock]] that just passes the source through. */
+  class ListenerLockWrapper(inner: ListenerLock, src: Async.Source[?]) extends ListenerLock:
+    val selfNumber: Long = inner.selfNumber
+    def lockSelf(_src: Async.Source[?]) =
+      inner.lockSelf(src)
+
+  /** Maps the lock of a listener, if it exists. */
+  inline def withLock[T](listener: Listener[?])(inline body: ListenerLock => T): T | Null =
+    listener.lock match
+      case null => null
+      case l: ListenerLock => body(l)
 
+  /** A simple [[Listener]] that always accepts the item and sends it to the consumer. */
   def acceptingListener[T](consumer: T => Unit) =
     new Listener[T]:
-      val topLock = null
+      val lock = null
       def complete(data: T, source: Source[T]) = consumer(data)
-      def release(to: ReleaseBoundary) = null
+      def release(to: LockMarker) = null
+
+  /** Returns a simple [[Listener]] that always accepts the item and sends it to the consumer. */
+  def apply[T](consumer: T => Unit): Listener[T] = acceptingListener(consumer)
```

Comment:
This is just an alias.
```suggestion
  inline def apply[T](consumer: T => Unit): Listener[T] = acceptingListener(consumer)
```